system/jlib/jevent.cpp

```diff
@@ -1757,6 +1821,513 @@ IEventVisitor* createDumpCSVEventVisitor(IBufferedSerialOutputStream& out)
     return new CDumpCSVEventVisitor(out);
 }
 
+class CEventFilter : public CInterfaceOf<IEventFilter>
+{
+protected:
+    class StringMatchHelper
+    {
+    private:
+        using Patterns = std::set<std::string>;
+        Patterns patterns;
+
+    public:
+        bool matches(const char* value) const
+        {
+            std::string tmp(value);
+            for (const std::string& pattern : patterns)
+            {
+                if (tmp.compare(pattern) == 0)
+                    return true;
+                if (isWildString(pattern.c_str()))
+                {
+                    if (WildMatch(value, pattern.c_str(), false))
+                        return true;
+                }
+                else
+                {
+                    try
+                    {
+                        RegExpr re(pattern.c_str(), false);
+                        if (re.find(value))
+                            return true;
+                    }
+                    catch (IException* e)
+                    {
+                        // Ignore exception and try next pattern
+                        e->Release();
+                    }
+                }
+            }
+            return false;
+        }
+
+        bool acceptToken(const char* token)
+        {
+            patterns.insert(token);
+            return true;
+        }
+    };
+
+    struct FilterTerm : public CInterface
+    {
+        virtual bool matches(EventType evt, const char* value) const { return true; }
+        virtual bool matches(EventType evt, bool value) const { return true; }
+        virtual bool matches(EventType evt, __uint64 value) const { return true; }
+        virtual bool accept(const char* values)
+        {
+            StringArray tokens;
+            tokens.appendList(values, ",", true);
+            ForEachItemIn(i, tokens)
+            {
+                if (!acceptToken(tokens.item(i)))
+                    return false;
+            }
+            return true;
+        }
+        virtual bool acceptToken(const char* token) = 0;
+    };
+
+    struct StringFilterTerm : public FilterTerm
+    {
+        StringMatchHelper helper;
+        bool matches(EventType evt, const char* value) const override
+        {
+            return helper.matches(value);
+        }
+
+        bool acceptToken(const char* token) override
+        {
+            return helper.acceptToken(token);
+        }
+    };
+
+    struct BoolFilterTerm : public FilterTerm
+    {
+        bool acceptTrue{false};
+        bool acceptFalse{false};
+
+        bool matches(EventType evt, bool value) const override
+        {
+            return (value && acceptTrue) || (!value && acceptFalse);
```

Comment:
picky: Maybe clearer as (value ? acceptTrue : acceptFalse)

You could also use an array[2]: acceptValue[value]
 - bool conversion to int is defined.